File: Medium/Java/Search_a_2D_Matrix(74).java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int start, end;
        int[] arr;
        for (int i=0; i<matrix.length; i++) {
            arr = matrix[i];
            start = arr[0];  // first and last elements
            end = arr[arr.length-1];
            if (start <= target && target <= end)  // if possible to be in arr
                return binarySearch(arr, target);  // search and return result
            
        }
        return false;
    }
    boolean binarySearch(int[] arr, int target) {
        int start=0, end=arr.length-1;
        int mid;
        
        while (start <= end)    {
            mid = (start+end) / 2;
            if (arr[mid] == target)
                return true;
            else if (arr[mid] < target)
                start = mid + 1;
            else
                end = mid - 1;
        }
        return false;  // not found
    }
}

File: Medium/C++/Search_a_2D_Matrix(74).cpp
class Solution {
public:
    
    bool searchMatrix(vector<vector<int>>& m, int target) {
        //double binary search
        if(!m.size() || !m[0].size()) return false;
        
        int rowThatProbablyHasTarget, mid;
        int lo = 0, hi = m.size()-1;
        
        
        while(lo < hi) {
            mid = (lo + hi)/2;
            
            if(m[mid].back() < target) {  //check if the greatest element of a row is smaller than the target
                lo = mid + 1;
            }
            
            else if(m[mid][0] > target) { 
                hi = mid-1;
            }
            else {
                lo = mid;
                break;
            }
        }
        
        
        rowThatProbablyHasTarget = lo;
        lo = 0;
        hi = m[rowThatProbablyHasTarget].size()-1;
        
        
        while (lo <= hi) {      //important case
            mid = (lo + hi)/2;
            
            if(m[rowThatProbablyHasTarget][mid] < target) {
                lo = mid + 1;
            }
            
            else if(m[rowThatProbablyHasTarget][mid] > target) {
                hi = mid-1;
            }
            
            else return true;
        }
        
        return false;
        
        
    }
};